Ilariae/delegated guardians (#812)
wormhole-foundation/wormhole-docs: docs/protocol/infrastructure/guardians.md

# Guardians

Wormhole relies on a set of 19 distributed nodes called Guardians that monitor the state on several blockchains. The current Guardian set can be seen in the [Dashboard](https://wormhole-foundation.github.io/wormhole-dashboard/#/?endpoint=Mainnet){target=\_blank}. Depending on the chain's operational configuration, messages may be observed directly by all Guardians or by a delegated subset of Guardians. Regardless of how observations are collected, all [Verifiable Action Approvals](/docs/protocol/infrastructure/vaas/){target=\_blank} (VAAs) are ultimately produced as standard 13-of-19 multisignature attestations, preserving full compatibility with existing contracts and integrators.

Guardians fulfill their role in the messaging protocol as follows:

1. Guardians observe messages and sign the corresponding payloads in isolation from the other Guardians.
    - On chains where all Guardians perform direct on-chain observation, each Guardian independently observes events directly.

>     - On chains configured for delegated observation, a subset performs direct observation and broadcasts a `DelegateObservation` to the rest of the network.

Reviewer: Focus nit: Prefer "delegated chain" over "chains configured for delegated observation":
```suggestion
    - On delegated chains, a subset performs direct observation and broadcasts a `DelegateObservation` to the rest of the network.
```

> 2. Once a sufficient delegated quorum is reached or once Guardians independently observe the event, signatures accumulate until the required 13-of-19 quorum is achieved.

Reviewer: See [naming convention](https://github.com/wormhole-foundation/wormhole-docs/pull/812/changes#r2842052376):
```suggestion
2. Once a sufficient delegate quorum is reached or once Guardians independently observe the event, signatures accumulate until the required 13-of-19 quorum is achieved.
```

3. This multisig represents proof that a majority of the Wormhole network has observed and agreed upon a state.

## Guardian Sets and Delegation

> The Guardian network comprises 19 Guardians. However, not all chains are secured in the same way. Some chains (such as Ethereum and Solana) are secured by all 19 Guardians. Each Guardian runs a full node and independently observes on-chain events. Other chains may be secured by a delegated subset of Guardians. Chains that are not explicitly configured for delegated observation default to full Guardian Set observation, meaning all 19 Guardians observe directly.

Reviewer: Focus nit: Prefer "delegated chain" over "configured for delegated observation"

For these chains:

- A configured delegated subset of the 19 Guardians performs direct on-chain observation.

> - Delegated Guardians broadcast a signed `DelegateObservation` gossip message.

Reviewer: ```suggestion
- Delegated Guardians broadcast a `SignedDelegateObservation` gossip message.
```

> - Canonical Guardians wait until a delegated quorum is reached before signing.

Reviewer: See [naming convention](https://github.com/wormhole-foundation/wormhole-docs/pull/812/changes#r2842052376):
```suggestion
- Canonical Guardians wait until a delegate quorum is reached before signing.
```

- A standard 13-of-19 VAA is ultimately produced.

This design provides operational flexibility while maintaining compatibility with the existing Wormhole contract stack.

### Delegated Observation Flow

> The following flow shows how a subset of Guardians directly observes events on those chains and broadcasts a signed `DelegateObservation` message over the Guardian gossip network. Canonical Guardians wait until the configured delegated quorum is reached before proceeding with the normal signing process. The final result is still a standard 13-of-19 VAA, fully compatible with existing smart contracts and integrations.

Reviewer: See [naming convention](https://github.com/wormhole-foundation/wormhole-docs/pull/812/changes#r2842052376):
```suggestion
The following flow shows how a subset of Guardians directly observes events on those chains and broadcasts a `SignedDelegateObservation` message over the Guardian gossip network. Canonical Guardians wait until the configured delegate quorum is reached before proceeding with the normal signing process. The final result is still a standard 13-of-19 VAA, fully compatible with existing smart contracts and integrations.
```

```mermaid
sequenceDiagram

participant CB as CoreBridge
participant DG as DelegatedGuardian
participant G as Gossip network
participant CG as CanonicalGuardian 1
participant CG2 as CanonicalGuardian 2

CB-->>DG: LogMessagePublished
DG->>DG: Watch & verify

alt DelegatedSet chain
  DG-->>G: Broadcast DelegateObservation
  G-->>CG: Receive DelegateObservation

>   G-->>CG2: Receive DelegateObservation

Reviewer: Optional: Depends on how pedantic we want to be.

```suggestion
  DG-->>G: Broadcast SignedDelegateObservation
  G-->>CG: Receive SignedDelegateObservation
  G-->>CG2: Receive SignedDelegateObservation
```

>   loop Wait for delegated quorum

Reviewer: See [naming convention](https://github.com/wormhole-foundation/wormhole-docs/pull/812/changes#r2842052376):
```suggestion
  loop Wait for delegate quorum
```

    CG->>CG: Mark message valid
    CG2->>CG2: Mark message valid
  end
end

note over CG,CG2: Standard signing flow continues
```

> ### Delegated Quorum Safeguards

Reviewer: See [naming convention](https://github.com/wormhole-foundation/wormhole-docs/pull/812/changes#r2842052376):
```suggestion
### Delegate Quorum Safeguards
```

> On chains configured for delegated observation, Canonical Guardians will not sign a message until a delegated quorum has been reached. This prevents a minority of Delegated Guardians from lowering the effective security threshold of a chain. For example, if a chain is configured as 7-of-9 Delegated Guardians, Canonical Guardians will only sign after at least 7 delegated observations agree. Once a delegated quorum is satisfied, Canonical Guardians sign to produce a standard 13-of-19 VAA.

Reviewer: Focus nit: Prefer "delegated chain" over "chains configured for delegated observation" and [naming convention](https://github.com/wormhole-foundation/wormhole-docs/pull/812/changes#r2842052376):
```suggestion
On delegated chains, Canonical Guardians will not sign a message until a delegate quorum has been reached. This prevents a minority of Delegated Guardians from lowering the effective security threshold of a chain. For example, if a chain is configured as 7-of-9 Delegated Guardians, Canonical Guardians will only sign after at least 7 delegate observations agree. Once a delegate quorum is satisfied, Canonical Guardians sign to produce a standard 13-of-19 VAA.
```

For these chains:

- A configured delegated subset of the 19 Guardians performs direct on-chain observation.
- Delegated Guardians broadcast a signed `DelegateObservation` gossip message.
- Canonical Guardians wait until a delegated quorum is reached before signing.
- A standard 13-of-19 VAA is ultimately produced.

This design provides operational flexibility while maintaining compatibility with the existing Wormhole contract stack.

### Delegated Observation Flow

The following flow shows how a subset of Guardians directly observes events on those chains and broadcasts a signed `DelegateObservation` message over the Guardian gossip network. Canonical Guardians wait until the configured delegated quorum is reached before proceeding with the normal signing process. The final result is still a standard 13-of-19 VAA, fully compatible with existing smart contracts and integrations.

```mermaid
sequenceDiagram

participant CB as CoreBridge
participant DG as DelegatedGuardian
participant G as Gossip network
participant CG as CanonicalGuardian 1
participant CG2 as CanonicalGuardian 2

CB-->>DG: LogMessagePublished
DG->>DG: Watch & verify

> note over DG: Wait for transfer verifier

Reviewer 2: We can remove this line from the user facing docs

To answer that, consider these key constraints and design decisions:

- **Threshold signatures allow flexibility, but**: With threshold signatures, in theory, any number of validators could participate. However, threshold signatures are not yet widely supported across blockchains. Verifying them is expensive and complex, especially in a chain-agnostic system.
- **t-Schnorr multisig is more practical**: Wormhole uses [t-Schnorr multisig](https://en.wikipedia.org/wiki/Schnorr_signature){target=\_blank}, which is broadly supported and relatively inexpensive to verify. However, verification costs scale linearly with the number of signers, so the size of the validator set needs to be carefully chosen.
- **19 Guardians form the canonical set**: A set of 19 participants represents a practical compromise between decentralization and efficiency. A quorum of 13 signatures is required to produce a valid VAA.

> - **Per-chain delegated thresholds**: On chains configured for delegated observation, a delegated subset of Guardians may be configured with a smaller observation threshold. Canonical Guardians wait for delegated quorum before contributing their signatures, ensuring that the effective security threshold for a chain cannot be reduced below its configured level.

Reviewer: Focus nit: Prefer "delegated chain" over "chains configured for delegated observation" and [naming convention](https://github.com/wormhole-foundation/wormhole-docs/pull/812/changes#r2842052376):
```suggestion
- **Per-chain delegated thresholds**: On delegated chains, a delegated subset of Guardians may be configured with a smaller observation threshold. Canonical Guardians wait for delegate quorum before contributing their signatures, ensuring that the effective security threshold for a chain cannot be reduced below its configured level.
```

Wormhole scales well, as demonstrated by its ability to handle substantial total value locked (TVL) and transaction volume even during tumultuous events.

On chains where all Guardians perform direct on-chain observation, each Guardian runs a full node and independently observes events. This requirement can be computationally heavy to set up; however, once all the full nodes are running, the Guardian Network's actual computation needs become lightweight.

> On chains configured for delegated observation, only the delegated subset of Guardians runs full nodes. Canonical Guardians rely on delegated observations broadcast through the gossip network and wait for delegated quorum before signing.

Reviewer: Focus nit: Prefer "delegated chain" over "chains configured for delegated observation" and [naming convention](https://github.com/wormhole-foundation/wormhole-docs/pull/812/changes#r2842052376):
```suggestion
On delegated chains, only the delegated subset of Guardians runs full nodes. Canonical Guardians rely on delegate observations broadcast through the gossip network and wait for delegate quorum before signing.
```
